cpp/src/io/utilities/datasource.cpp

```diff
@@ -199,6 +200,65 @@ class direct_read_source : public file_source {
   }
 };
 
+/**
+ * @brief Implementation class for reading from a device buffer source
+ */
+class device_buffer_source final : public datasource {
+ public:
+  explicit device_buffer_source(cudf::device_span<std::byte const> d_buffer) : _d_buffer{d_buffer}
+  {
+  }
+
+  size_t host_read(size_t offset, size_t size, uint8_t* dst) override
+  {
+    auto const count = std::min(size, this->size() - offset);
+    CUDF_CUDA_TRY(cudaMemcpy(dst, _d_buffer.data() + offset, count, cudaMemcpyDefault));
+    return count;
+  }
+
+  std::unique_ptr<buffer> host_read(size_t offset, size_t size) override
+  {
+    auto const count = std::min(size, this->size() - offset);
+    std::vector<std::byte> h_data(count);
+    CUDF_CUDA_TRY(cudaMemcpy(h_data.data(), _d_buffer.data() + offset, count, cudaMemcpyDefault));
+    return std::make_unique<owning_buffer<std::vector<std::byte>>>(std::move(h_data));
+  }
+
+  [[nodiscard]] bool supports_device_read() const override { return true; }
+
+  std::future<size_t> device_read_async(size_t offset,
+                                        size_t size,
+                                        uint8_t* dst,
+                                        rmm::cuda_stream_view stream) override
+  {
+    auto const count = std::min(size, this->size() - offset);
+    CUDF_CUDA_TRY(
+      cudaMemcpyAsync(dst, _d_buffer.data() + offset, count, cudaMemcpyDefault, stream.value()));
+    return std::async(std::launch::deferred, [count] { return count; });
```

Comment:
How do we know this future is stream-synced when calling `.get()`? If that doesn't matter, then why use a future?

Reply:
The naming is a bit misleading. The "async" here refers to the fact that device reads are launched asynchronously from the host side. Other data sources seem to launch multiple device reads (operations more comlplex than just invoking `cudaMemcpy`) and then sync to get the final data. For the case of device buffer, this operation is one single invocation of `cudaMemcpy` so we don't really need the `std::future`. Returning a future is not necessary in this case and we are doing so simply because the base virtual function we are overriding is defined in this way.

Reply:
Because GDS did not have an async API at the time of our integration (they do now, but it's not stream-ordered), we have odd semantics of the async reads. Async here means "once the future returns, data is available in the given stream". Sometimes the future part does not matter (like here), and sometimes the stream part might not matter. But calling read_async(..., stream), then get() on the returned future, then stream.synchronize() on the passed stream should always mean that the data is at the destination from that point forward.